src/Raven.Server/RavenServer.cs

```diff
@@ -929,45 +938,179 @@ internal AuthenticateConnection AuthenticateConnectionCertificate(X509Certificat
+            // with the same public key pinning hash.
+            var pinningHash = certificate.GetPublicKeyPinningHash();
+            var certsWithSameHash = ServerStore.Cluster.GetCertificatesByPinningHashSortedByExpiration(ctx, pinningHash);
+
+            if (certsWithSameHash.Count > 0)
+            {
+                // Hash is good, let's validate it was signed by a known issuer, otherwise users can use the private key to register a new cert with a different issuer.
+                var goodKnownCert = new X509Certificate2(Convert.FromBase64String(certsWithSameHash[0].Certificate));
+                if (CertHasKnownIssuer(certificate, goodKnownCert, out var issuerHash) == false)
+                {
+                    if (_authAuditLog.IsInfoEnabled)
+                        _authAuditLog.Info($"Connection from {remoteAddress} with certificate '{certificate.Subject} ({certificate.Thumbprint})'. " +
+                                          "The client certificate is not registered in the cluster explicitly but is trusted implicitly by its Public Key Pinning Hash. " +
+                                          "The client certificate was signed by an unknown issuer - closing the connection. " +
+                                          $"To fix this, the admin can register the pinning hash of the *issuer* certificate: '{issuerHash}' in the 'Security.WellKnownIssuerHashes.Admin' configuration entry.");
+
+                    authenticationStatus.Status = AuthenticationStatus.UnfamiliarIssuer;
+                    authenticationStatus.IssuerHash = issuerHash;
+                    return;
+                }
+
+                // Success, we'll add the new certificate with same permissions as the original
+                var existingCert = certsWithSameHash[0];
+                var newCertBytes = certificate.Export(X509ContentType.Cert);
+
+                var newCertDef = new CertificateDefinition()
+                {
+                    Name = existingCert.Name,
+                    Certificate = Convert.ToBase64String(newCertBytes),
+                    Permissions = existingCert.Permissions,
+                    SecurityClearance = existingCert.SecurityClearance,
+                    Password = existingCert.Password,
+                    Thumbprint = certificate.Thumbprint,
+                    PublicKeyPinningHash = pinningHash,
+                    NotAfter = certificate.NotAfter
+                };
+
+                cert = ServerStore.Cluster.GetCertificateByThumbprint(ctx, certificate.Thumbprint) ??
+                           ServerStore.Cluster.GetLocalStateByThumbprint(ctx, certificate.Thumbprint);
+                if (cert != null)
+                    return;
+
+                // This command will discard leftover certificates after the new certificate is saved.
+                ServerStore.SendToLeaderAsync(new PutCertificateWithSamePinningHashCommand(certificate.Thumbprint, newCertDef));
```

Comment:
Don't we get a warning here about ignoring the return value?

Reply:
Fixed